### DISCUSSION TITLE
Strange behaviour while using RUN_COMMAND Intent

### DISCUSSION THREAD

**Discussion by coderGtm** [termux/termux-app#3640] [Q&A] [answered]
Hello

_This is my first post here and I would like to narrate a strange incident..._

I am the developer of an app which is a Terminal App Launcher. You can view it on GitHub here: https://github.com/coderGtm/yantra-app-launcher. What's relevant is that it has a `termux` command that lets users execute termux commands directly from my app, like `termux top -n 5` executes `top -n 5` in termux using the Termux Run Command Intent.

But everytime Termux was not in the background or not opened recently, the command failed with error which stated 

Unable to start service Intent for RUN_COMMAND. App is in background

Recently, I enabled wakelock for a termux session and tried using the `termux` command from my app, while termux was not in background, and the command worked perfectly. After this, I released the Wakelock.

Fast forward 2 days, I did not open termux in these 2 days even once. Now when I used the termux command with **no wakelock and app also not in background** for 2 days, the command works perfectly...**not once, not twice but everytime....**

I failed to understand the mechanism behind this strange incidents. I guess it's some trust issues with Android and Termux, as when I once granted wakelock permission to termux, Android started trusting termux and let it open from background.

I am asking this because I need to give a proper solution to users of my app who want to use termux also from my app but it's too hard to set up for them with such kind of strange issues!

**Comment by agnostic-apollo** [comment]
It might be related to https://github.com/termux/termux-app/issues/3338#issuecomment-1505591761 You can try `startForegroundService()`.

https://github.com/coderGtm/yantra-app-launcher/blob/3850a2160373d4a4ef81ed66cf6bc71809f615a0/app/src/main/java/com/coderGtm/yantra/activities/MainActivity.kt#L1719

If termux app was started two days ago and its foreground service notification remained, then it will still be considered a foreground app. Also foreground state might stay for a while after app is moved to background if it was opened. Force close termux app first to do testing.

**Reply by coderGtm** [reply]
The foreground service notification did not remain. Even today, again after 2 days of asking the question, I tried the `termux` command and it worked. I even force-stopped the app right now and then tried the command, still works!

I don't know how to revoke the permission for wake lock. Can you please tell me? After revoking the permission, I'll test again to see the results. If it does not work then I can use `startForegroundService()` to solve it.

Thanks

**Reply by agnostic-apollo** [reply, marked as answer]
Ah, when you enabled a wakelock in termux, it would have asked to be exempted from battery optimizations (`REQUEST_IGNORE_BATTERY_OPTIMIZATIONS`), that would be why error is not triggering any more. Draw over apps permission should have same effect. Could revoke it from `Android Settings` -> `Apps` -> `Termux` -> `Advanced` -> `Battery` -> `Battery optimizations` -> `Termux` -> `Don't optimize`. Wakelock permission (`WAKE_LOCK`) itself is automatically granted.

https://developer.android.com/guide/components/foreground-services#background-start-restriction-exemptions

**Reply by agnostic-apollo** [reply]
Disabling battery optimizations for termux is also documented.

https://github.com/termux/termux-app/wiki/RUN_COMMAND-Intent#battery-optimizations-may-be-mandatory-depending-on-device
